### DISCUSSION TITLE
User control over installing dependencies

### DISCUSSION THREAD

**Discussion by kbrandwijk** [StjerneIdioten/GE-Blender-Tools#2] [Archived]
The Blender Foundation has strong opinions about installing dependencies without explicit user consent. The addon currently tries to silently install lxml if it happens to have enough permissions.

I would suggest adopting a more user-centric pattern like the one in this example repo: https://github.com/robertguetzkow/blender-python-examples/tree/master/add_ons/install_dependencies

**Comment by StjerneIdioten** [comment]
It does do it without receiving explicit permission more out of necessity than anything at the time :-P there is plenty of log output about what's going on though. Which is more than can be said for a lot of addons requiring external dependencies. The foundation might have strong opinions about it, but it's also been unnecessarily hard to manage for addon develops and until recently you had to use all sorts of hacks to even get it working and there is still plenty of opportunity to clash between addons and their dependencies.

But I agree with it being a button instead, to specifically give permission. Will think of it for the next iteration.

**Reply by kbrandwijk** [reply]
Oh I feel your pain, I'm just happy that a kind of standard approach is emerging out of it that is both user and developer friendly.

**Comment by StjerneIdioten** [comment]
![image](https://user-images.githubusercontent.com/11218511/149052838-5c113af8-9200-4c29-9a7b-66de4737ba86.png)

I have chosen to implement a solution based on the link you sent, with some modifications. I've gone with a list overview off all the external dependencies and the currently registered versions vs. the needed versions and the ability to individually install/upgrade libraries.

I still need to put in all the warnings, hints about elevating to administrative and so on as well as making the "upgrade" actually do an upgrade. I don't count on the addon having a lot of dependencies though apart from the GE-Core library, lxml isn't even necessary anymore and I just have it there for testing.

I chose to do it this way because you might run into a situation where the addon requires a library that is already installed with another version for another addon and I wanted to give people the possibility of just upgrading one package at a time to see if something potentially breaks, which is very much a possibility with this wild west of installing packages 😉 

The dependency handling in code is quite straight forward, currently it's a tuple of "dependencies" just like the suggested solution, just with the version number added as well. The UI and everything set's itself up from this tuple, so it is very easy to add further dependencies. It could potentially even just be moved into a configuration file and integrate with existing python tools for dependency management (To automate it with CI) but that's probably overkill for the dependencies needed here 😉 

![image](https://user-images.githubusercontent.com/11218511/149054004-b133d0ba-65e6-494f-b37f-8917a5454aa0.png)

**Reply by StjerneIdioten** [reply]
This has now been addressed in 251f0b5bf036075443987c02d6d617443d9bc953 and added, a description of the functionality will be added to the docs once I have those up and running.
